protocols/kad/src/behaviour.rs

```diff
@@ -315,134 +337,206 @@ where
         self.kbuckets.iter().map(|entry| entry.node.key.preimage())
     }
 
-    /// Starts an iterative `FIND_NODE` request.
+    /// Performs a lookup for the closest peers to the given key.
     ///
-    /// This will eventually produce an event containing the nodes of the DHT closest to the
-    /// requested `PeerId`.
-    pub fn find_node(&mut self, peer_id: PeerId) {
-        self.start_query(QueryInfoInner::FindPeer(peer_id));
-    }
-
-    /// Starts an iterative `GET_PROVIDERS` request.
-    pub fn get_providers(&mut self, target: Multihash) {
-        self.start_query(QueryInfoInner::GetProviders { target, pending_results: Vec::new() });
+    /// The result of this operation is delivered in [`KademliaEvent::GetClosestPeersResult`].
+    pub fn get_closest_peers<K>(&mut self, key: K)
+    where
+        K: Into<Multihash> + Clone
+    {
+        let multihash = key.into();
+        let info = QueryInfo::GetClosestPeers { key: multihash.clone() };
+        let target = kbucket::Key::new(multihash);
+        let peers = self.kbuckets.closest_keys(&target);
+        let inner = QueryInner::new(info);
+        self.queries.add_iter_closest(target.clone(), peers, inner);
     }
 
-    /// Starts an iterative `GET_VALUE` request.
+    /// Performs a lookup for a record in the DHT.
     ///
-    /// Returns a number of results that is in the interval [1, 20],
-    /// if the user requested a larger amount of results it is cropped to 20.
-    pub fn get_value(&mut self, key: &Multihash, num_results: NonZeroU8) {
-        let num_results = usize::min(num_results.get() as usize, kbucket::MAX_NODES_PER_BUCKET);
-        let mut results = Vec::with_capacity(num_results);
-
-        if let Some(record) = self.records.get(key) {
-            results.push(record.into_owned());
-            if num_results == 1 {
-                self.queued_events.push(NetworkBehaviourAction::GenerateEvent(
-                    KademliaOut::GetValueResult(
-                        GetValueResult::Found { results }
-                )));
-                return;
+    /// The result of this operation is delivered in [`KademliaEvent::GetRecordResult`].
+    pub fn get_record(&mut self, key: &Multihash, quorum: Quorum) {
```

Comment:
I think that we shouldn't be handing out to users the `Multihash` types in the api, since it would make a lot of code in substrate depentable on `Multihash` and different places in code can hash things differently and so on. What if the api took bytes and did the hashing?

Reply:
Choosing multihashes as record keys was done in your original PR and didn't change here. I think it would be fine (even good) to further generalise over the record key type, but that should be done in another PR - this one is big enough as it is.

> What if the api took bytes and did the hashing?

Not sure what you mean by that. Whatever the key is, it is _always_ hashed into the Kademlia keyspace, see `protocols/kad/kbucket/key.rs`.

Reply:
For what it's worth, I have a branch [here](https://github.com/romanb/rust-libp2p/compare/kad-records-providers...romanb:kad-key-bytes?expand=1) that sits on top of this PR and generalises record keys. I may open a PR for that once / if this PR is merged.